### PR TITLE
TFA Flow

### LINKED ISSUE #1204: Implement TFA enable
Pages / steps:

* [ ] Enable
* [ ] Confirm
* [ ] Confirmed

### PR DESCRIPTION
**Pull request by na3d**
Closes #1204

A cleaned up branch of PR [1525](https://github.com/linode/manager/pull/1526)

### CODE REVIEW

File: components/modals/ConfirmModalBody.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ export default class ConfirmModalBody extends Component {
     const { loading } = this.state;
     return (
       <div className={`ConfirmModalBody-body ${className}`}>
-        {React.isValidElement(children) ? children : <p>{children}</p>}
+        {React.isValidElement(children) ? children : <div>{children}</div>}
```
Comment (eatonphil) [inline components/modals/ConfirmModalBody.js]: Why the change? The p gave it padding it needed.

File: src/profile/components/TwoFactorModal.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+  twoFactorScratchModal(scratch) {
+    const { dispatch } = this.props;
+    dispatch(showModal('Scratch code generated',
+      <ConfirmModalBody
+        buttonText="Ok"
+        onOk={() => dispatch(hideModal())}
+        onCancel={() => dispatch(hideModal())}
+      >
+        <div className="form-group">
```
Comment (eatonphil) [inline src/profile/components/TwoFactorModal.js]: Please use FormGroup instead of being explicit here.

File: test/profile/components/TwoFactorPanel.spec.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import sinon from 'sinon';
+import { shallow } from 'enzyme';
+import { expect } from 'chai';
+import { profile } from '@/data/profile';
+
+import { expectRequest } from '@/common';
+import { TwoFactorPanel } from '~/profile/components/TwoFactorPanel';
+
+describe('profile/layouts/twofactorpanel', () => {
```
Comment (eatonphil) [inline test/profile/components/TwoFactorPanel.spec.js]: Correct capitalization please?

File: src/profile/components/TwoFactorPanel.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        this.twoFactorModal(tfaResponse);
+      }
+    } catch (response) {
+      if (!response.json) {
+        // eslint-disable-next-line no-console
+        return console.error(response);
+      }
+
+      const errors = await reduceErrors(response);
+      this.setState({ errors: errors });
```
Comment (eatonphil) [inline src/profile/components/TwoFactorPanel.js]: You should use dispatchOrStoreErrors here instead. grep for that for examples.

File: src/profile/components/TwoFactorModal.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+          </div>
+          <div className="form-group row">
+            <div className="col-sm-12">
+              <div className="form-group">
+                If your TFA app does not have a QR scanner, you can use this secret key.
+              </div>
+              <div className="alert alert-warning">{secret}</div>
+            </div>
+          </div>
+          <div className="form-group">
```
Comment (eatonphil) [inline src/profile/components/TwoFactorModal.js]: Again, you should use FormGroup here instead.

File: src/profile/components/TwoFactorModal.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+      toggleTwoFactor();
+      this.twoFactorScratchModal(tfaResponse.scratch);
+    } catch (response) {
+      if (!response.json) {
+        // eslint-disable-next-line no-console
+        return console.error(response);
+      }
+
+      const errors = await reduceErrors(response);
+      this.setState({ errors: errors });
```
Comment (eatonphil) [inline src/profile/components/TwoFactorModal.js]: You should use dispatchOrStoreErrors here instead of all of this.

File: components/modals/ConfirmModalBody.js (hunk 1 above)
Comment (abemassry) [inline components/modals/ConfirmModalBody.js]: There were warnings that you couldn't nest a <div> inside a <p>

Comment (eatonphil) [inline components/modals/ConfirmModalBody.js]: Then the terniary operator needs to be modified, not the tag here.

File: src/profile/components/TwoFactorModal.js (hunk 2 above)
Comment (na3d) [inline src/profile/components/TwoFactorModal.js]: ✅

File: test/profile/components/TwoFactorPanel.spec.js (hunk 3 above)
Comment (na3d) [inline test/profile/components/TwoFactorPanel.spec.js]: ✅

File: src/profile/components/TwoFactorModal.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+          onOk={async () => {
+            await this.twoFactorConfirm();
+            if (errors === {}) {
+              dispatch(hideModal());
+            }
+          }}
+          onCancel={() => dispatch(hideModal())}
+        >
+          <div className="form-group row">
+            <div className="col-sm-12">
```
Comment (eatonphil) [inline src/profile/components/TwoFactorModal.js]: You can just drop the row and the col-sm-12 here.

File: src/profile/components/TwoFactorModal.js (hunk 5 above)
Comment (na3d) [inline src/profile/components/TwoFactorModal.js]: ✅

File: src/profile/components/TwoFactorModal.js (hunk 7 above)
Comment (na3d) [inline src/profile/components/TwoFactorModal.js]: ✅
